### PR TITLE
Allow specification of file object in sec expression

### PR DESCRIPTION
**Pull request by illusional**
Fixes: common-workflow-language/cwltool/issues/1232

If the type returned by the expression is an object, then the code should set the values of `sf_location` and `sf_name` rather than derive them from an expected string value.

### CODE REVIEW

**Comment by illusional** [comment]
Hey @mr-c, any advice for this PR?

**Comment by mr-c** [comment]
@illusional I replied on Discourse

**Comment by illusional** [comment]
Hi @mr-c, I [replied](https://cwl.discourse.group/t/ask-cwl-to-rename-a-secondary-file/72/4?u=illusional) to your discourse message, and feel that this PR is still relevant, fixes something that the spec says should be supported, and was your recommendation in [#1229](https://github.com/common-workflow-language/cwltool/issues/1229#issuecomment-566447158).

**Comment by tetron** [comment]
@illusional I responded on the discourse site, I think the fact that you found a way to crash cwltool shows this isn't a fully thought out code path, but we need to consider what the desired behavior is, document it, and add tests (probably cwl 1.2 conformance tests) for the correct behavior.

**Comment by illusional** [comment]
Hey @mr-c / @tetron, 

As per the Pacific-friendly CWL call and per your suggestion on the discourse, I've prepared two separate conformance tests. I'm not totally sure where to add them (maybe back in v1.0). 

I've also replied to your concerns on the discourse.

## rename-inputs.cwl

Required files:
```bash
touch secondary_file_test.txt
touch secondary_file_test.txt.accessory
```

Job: **rename-inputs.json**
```yaml
inputWithSecondary: secondary_file_test.txt
```

Conformance Test:
```yaml
- job: v1.0/secondary_file_test.json
  tool: v1.0/rename-inputs.cwl
  doc: Confirm CommandInputParameter expression can receive a File object
  output:
    output_file:
      class: File
      checksum: sha1$901c3d387a263c57eaed6f24a82517c1fb0e198d
      size: 54
      location: Any
  tags: [ inline_javascript, secondary_files ]
```

<details><summary>Click to show <code>rename-inputs.cwl</code></summary><p>

```cwl
#!/usr/bin/env cwl-runner
id: InputSecondaryFileConformanceTest
baseCommand:
- ls
class: CommandLineTool
cwlVersion: v1.0
doc: |
  Simple test to confirm the implementation of expressions returning a File within a CommandInputParameter.secondaryFile field.

  Use GREP to filter the result from ls to ensure we only get the secondary files in there.
  
  Related links:
  - Issue: https://github.com/common-workflow-language/cwltool/issues/1232
  - PR: https://github.com/common-workflow-language/cwltool/pull/1233
  - Discourse: https://cwl.discourse.group/t/ask-cwl-to-rename-a-secondary-file/72

inputs:
- id: inputWithSecondary
  type: File
  doc: |
    This input will with a secondary file `.accessory`. You could create these files (and its accessory) with:
    ```bash
    touch secondary_file_test.txt
    touch secondary_file_test.txt.accessory
    ```
  secondaryFiles: |
    ${
        function resolveSecondary(base, secPattern) {
          if (secPattern[0] == '^') {
            var spl = base.split('.');
            var endIndex = spl.length > 1 ? spl.length - 1 : 1;
            return resolveSecondary(spl.slice(undefined, endIndex).join("."), secPattern.slice(1));
          }
          return base + secPattern;
        }
        return [{
            location: resolveSecondary(self.location, '.accessory'),
            basename: resolveSecondary(self.basename, '^.accessory')
        }];
    }

arguments:
- valueFrom: "|"
  shellQuote: false
  position: 0
- valueFrom: "grep"
  position: 1
- valueFrom: "secondary"
  position: 2

outputs:
- id: output_file
  type: stdout

requirements:
  InlineJavascriptRequirement: {}
  ShellCommandRequirement: {}
  InitialWorkDirRequirement:
    listing:
    - $(inputs.inputWithSecondary)
```
</p></details>

## rename-outputs.cwl

_No required files or jobs_

Conformance test:
```yaml
- job: v1.0/blank.json # or can I just leave this field blank
  tool: v1.0/rename-outputs.cwl
  doc: Confirm CommandOutputParameter expression can receive a File object
  output:
    output_file:
      class: File
      checksum: sha1$da39a3ee5e6b4b0d3255bfef95601890afd80709
      size: 0
      location: Any
      secondaryFiles:
      - class: File
        checksum: sha1$da39a3ee5e6b4b0d3255bfef95601890afd80709
        location: Any
        size: 0
  tags: [ inline_javascript, secondary_files ]
```


<details><summary>Click to show <code>rename-outputs.cwl</code></summary><p>

```cwl
#!/usr/bin/env cwl-runner
id: OutputSecondaryFileConformanceTest
baseCommand:
- ls
class: CommandLineTool
cwlVersion: v1.0
doc: |
  Simple test to confirm the implementation of expressions returning a File within a CommandOutputParameter.secondaryFile field.

  Related links:
  - Issue: https://github.com/common-workflow-language/cwltool/issues/1232
  - PR: https://github.com/common-workflow-language/cwltool/pull/1233
  - Discourse: https://cwl.discourse.group/t/ask-cwl-to-rename-a-secondary-file/72

inputs: []
outputs:
- id: output_file
  label: out
  outputBinding:
    glob: "*.txt"
  secondaryFiles: |
   ${
        function resolveSecondary(base, secPattern) {
          if (secPattern[0] == "^") {
            var spl = base.split(".");
            var endIndex = spl.length > 1 ? spl.length - 1 : 1;
            return resolveSecondary(spl.slice(undefined, endIndex).join("."), secPattern.slice(1));
          }
          return base + secPattern;
        }
        return [
          {
            path: resolveSecondary(self.path, "^.accessory"),
            basename: resolveSecondary(self.basename, ".accessory")
          }
        ];
   }
  type: File
requirements:
  InlineJavascriptRequirement: {}
  InitialWorkDirRequirement:
    listing:
    - entry: ""
      entryname: secondary_file_test.txt
    - entry: ""
      entryname: secondary_file_test.accessory
```
</p></details>

**Comment by illusional** [comment]
Hey @tetron, what's the follow-up process for me here?

I've now included conformance tests and issued a PR against CWL v1.2 (https://github.com/common-workflow-language/cwl-v1.2/pull/18).

Is this PR right to merge, do I need to get the conformance tests merged first?

File: cwltool/builder.py [hunk 1]
```diff
@@ -398,12 +405,6 @@ def _capture_files(f):  # type: (Dict[str, str]) -> Dict[str, str]
                                 if d["basename"] == sfname:
                                     found = True
                             if not found:
-                                sf_location = (
-                                    datum["location"][
-                                        0 : datum["location"].rindex("/") + 1
-                                    ]
-                                    + sfname
-                                )
                                 if isinstance(sfname, MutableMapping):
```
[the remark this answers is not in the document]
Comment (illusional) [reply cwltool/builder.py:418]: Do you want me to remove the lines:

```python
if isinstance(sfname, MutableMapping):
    datum["secondaryFiles"].append(sfname)
```

It's not strictly part of this PR, but I agree that `sfname` should always be a string by this point.

File: cwltool/builder.py [hunk 2]
```diff
@@ -390,6 +390,13 @@ def _capture_files(f):  # type: (Dict[str, str]) -> Dict[str, str]
                             if not sfname:
                                 continue
                             found = False
+
+                            if isinstance(sfname, str):
+                                sf_location = datum["location"][0:datum["location"].rindex("/")+1]+sfname
+                            else:
```
[the remark this answers is not in the document]
Comment (illusional) [reply cwltool/builder.py]: Changed this to now be:

```python
elif isinstance(sfname, MutableMapping):
    sf_location = sfname["location"]
    sfname = sfname["basename"]
else:
    raise WorkflowException(
        "Expected secondaryFile expression to return type 'str' or 'MutableMapping', received '%s'" % (type(sfname))
    )
```

**Comment by illusional** [comment]
Hi @tetron, thanks for taking the time to review this. I've made one change per your request and commented on your other. I've confirmed that my conformance tests still run.

**Comment by mr-c** [comment]
This needs testing against https://github.com/common-workflow-language/cwl-v1.2/pull/18/commits/dbfa31a97ccefb773471740a09da41658d917b1e

**Comment by tetron** [comment]
I made a few more changes to handle another edge case and also to reflect some tweaks I made to the conformance test.

**Comment by illusional** [comment]
Thanks @tetron, these look good. I looked through the [TravisCI build](https://travis-ci.org/github/common-workflow-language/cwltool/builds/674964440) and it didn't seem to have the new tests in there. The failures seemed maybe transient, and might be fixed by a re-test - I don't have access to trigger this though.

**Comment by tetron** [comment]
@illusional there's some kind of dependency version conflict happening, also known as "some dude on the internet broke our build".  It probably just requires adjusting a version pin somewhere, the trick is figuring out where.
